design/expected-hardware-configuration-validation.md

```diff
@@ -0,0 +1,138 @@
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints ](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+      * [References](#references)
+
+## Summary
+
+Code with the ability to perform the comparison on inspected and ready hosts and update the host with labels containing matched profiles.
+
+## Motivation
+
+We are closely working with Airship team, There is one requirement in Airship for classification.
+As an User, I want to validate and classify the hosts based on my hardware requirement so that I will get to know how many hosts match to my profile and accordingly I can go ahead and provision these hosts only.
+This feature helps user to avoid runtime failures and also increases the performances for workload deployments.
+
+### Goals
+
+Automatically classify bare metal hosts based on user-provided rules in a way that other controllers can use, without having to evaluate all of those rules themselves.
+
+## Proposal
+
+We are creating proposal based on the discussion with metal3 community on issue #351:
+https://github.com/metal3-io/baremetal-operator/issues/351
+
+We compared introspection data of Ironic with metal3 schema and we found that default  hardware
+configuration needs to be added by introducing a new CRD `HardwareClassification`.
+
+We will write a CRD that checks inspected baremetal hosts against hardware configuration and add label as "MATCHES" for matched profile into baremetal host.
+  
+### Implementation Details/Notes/Constraints
+
+Link for Existing Metal3 Specs
+Please refer metal3 spec for bare-metal:
+https://github.com/metal3-io/baremetal-operator/blob/master/deploy/crds/metal3.io_baremetalhosts_crd.yaml
+
+* Below is sample yaml of Kind HardwareClassification.
+    
+   ```yaml
+   
+   apiVersion: metal3.io.sigs.k8s.io/v1alpha1
+   kind: HardwareClassification
+   metadata:
+     name: profile1
+   spec:
+   expectedValidationConfiguration:
+      namespace: metal3
+      CPU:
+         minimumCount: 4
+         minimumSpeed: 4.3
+      Disk:
+         minimumSizeGB: 2200
+         minimumCount: 2
+      NICS:
+         minimumcount: 4
+      RAM:
+         minimumSizeGB: 32
+   ```
+   Similarly, user can mention maximum requirements for CPU, Disk, NICS and RAM.
+
+* Write a new API as hardware-classification/api/v1alpha1 and new kind(CRD) HardwareClassification.
+    
+    - Create the Schema struct for `ExpectedHardwareConfiguration` inside `HardwareClassificationSpec`,
+    in file /api/v1alpha1/hardwareClassification_types.go.
+
+    - Implement a new function fetchHost() which will fetch all baremetal hosts from Baremetal-Operator of namespace 'metal3' provided by user in CRD_yaml. We will add a watch on the baremetalhost kind for the hardware setting changes.
```

Comment:
There is work going on (maybe it has landed) to make the baremetal-operator optionally able to look at all namespaces. I think we will need this controller to work the same way.

Reply:
okay. I will discuss about it with you to get more details on it.

Reply:
https://github.com/metal3-io/baremetal-operator/pull/471

Reply:
When we discussed this, we decided to start that the classification resource would be namespace scoped and only apply to hosts in the same namespace.